pkg/util/testutil/patternscanner.go

```diff
@@ -15,32 +15,49 @@ import (
 	"testing"
 )
 
+// NoPattern is a sugar syntax for empty pattern
+var NoPattern *regexp.Regexp
+
 // PatternScanner is a helper to scan logs for a given pattern.
 type PatternScanner struct {
-	// The log pattern to match on
-	pattern *regexp.Regexp
+	// The log pattern to match on validating successful start
+	startPattern *regexp.Regexp
+	// The log pattern to match on validating successful finish. This is optional
+	finishPattern *regexp.Regexp
 	// Once we've found the correct log, we should notify the caller.
 	DoneChan chan struct{}
 	// A sync.Once instance to ensure we notify the caller only once, and stop the operation.
 	stopOnce sync.Once
-	// A helper to spare redundant calls to the analyzer once we've found the relevant log.
+
+	// flag to indicate that start was found, and we should look for finishPattern now
+	started bool
+	// A helper to spare redundant calls to the analyzer once we've found both start and finish
 	stopped bool
 
 	// keep the stdout/err in case of failure
 	buffers []string
-
 	//Buffer for accumulating partial lines
 	lineBuf string
 }
 
 // NewScanner returns a new instance of PatternScanner.
-func NewScanner(pattern *regexp.Regexp, doneChan chan struct{}) *PatternScanner {
-	return &PatternScanner{
-		pattern:  pattern,
-		DoneChan: doneChan,
-		stopOnce: sync.Once{},
-		stopped:  false,
+// startPattern and finishPattern are both optional, if none provided, Done channel will be signaled immediately.
+func NewScanner(startPattern, finishPattern *regexp.Regexp, doneChan chan struct{}) *PatternScanner {
+	ps := &PatternScanner{
+		startPattern:  startPattern,
+		finishPattern: finishPattern,
+		DoneChan:      doneChan,
+		stopOnce:      sync.Once{},
+		// skip looking for start pattern if not provided
+		started: startPattern == nil,
+		// mark the scanner as stopped if no pattern was provided
+		stopped: startPattern == nil && finishPattern == nil,
```

Comment:
wondering what is the usage of the scanner in such a case
since to me like returning an error is more appropriate

The current implementation feels hacky

Reply:
originally i was thinking to leave an option just to record logs (btw current implementation doesn't do it), but i think you are right, the name of this struct suggests a different functionality rather than redirecting logs. I will change the creation logic and make the `startPattern` required